EventGenerator/src/RPCGun_module.cc

```diff
@@ -56,214 +54,213 @@ namespace mu2e {
+            while(part->parent().isNonnull()) {
+              if((part->creationCode() == ProcessCode::mu2ePrimary)) { //TODO - do we need this?
+                part = part->parent();
+                std::cout<<" while primary "<<part->creationCode()<<" "<<part->pdgId()<<std::endl;
+                if(part->pdgId() != 2212){
+                   tau += part->endProperTime() / gc.getParticleLifetime(part->pdgId());
+                   std::cout<<"adding time "<<tau<<std::endl;
+                 }
+              }else {
+                part = part->parent();
+                if ( std::binary_search(decayOffPDGCodes_.begin(), decayOffPDGCodes_.end(), int(part->pdgId()) ) ) { //is part in the list?
+                  tau += part->endProperTime() / gc.getParticleLifetime(part->pdgId()); 
+                  std::cout<<"in else part->parent() code "<<part->creationCode()<<" "<<part->pdgId()<<std::endl;
+                  std::cout<<"adding time "<<tau<<std::endl;
+                }
             }
-          }else{
-            stop = tstop;
-            break;
-          }
-        }
-      }
-    }else{
-      timemap->SetTime(protonPulse_->fire());
-      if (applySurvivalProbability_){
-        while (true){
-          const auto& tstop = stops_.fire();
-          double weight = exp(-tstop.tauNormalized)*survivalProbScaling_;
-          double rand = randomFlat_.fire();
-          if (weight > rand){
-            stop = tstop;
-            break;
-          }
+          } 
         }
-      }else{
-        const auto& tstop = stops_.fire();
-        stop = tstop;
       }
-    }
-
-    //std::cout << "Found stop " << exp(-stop.tauNormalized) << " " << stop.t << std::endl;
-
-    const CLHEP::Hep3Vector pos(stop.x, stop.y, stop.z);
-
-    const double energy = generateEnergy();
+   
+    weight = exp(-tau);
+    if(doHistograms_) _htime->Fill(weight);
+    return weight;
+  }
 
-    if(!generateInternalConversion_){
-      output->emplace_back( PDGCode::gamma,
-                            GenId::ExternalRPC,
-                            pos,
-                            CLHEP::HepLorentzVector( randomUnitSphere_.fire(energy), energy),
-                            stop.t );
+  //================================================================
+  void RPCGun::produce(art::Event& event) {
+    auto output{std::make_unique<StageParticleCollection>()};
+    const auto simh = event.getValidHandle<SimParticleCollection>(simsToken_); 
+    const auto pis = stoppedPiMinusList(simh);
+    double time_weight = MakeEventWeight(event);
+    std::unique_ptr<EventWeight> pw(new EventWeight(time_weight));
+    event.put(std::move(pw)); 
+    if(pis.empty()) {
+      throw   cet::exception("BADINPUT")
+        <<"RPCGun::produce(): no suitable stopped pion in the input SimParticleCollection\n";
+    }
 
-      event.put(std::move(output));
-      event.put(std::move(timemap));
-    } else {
+    unsigned int randIn = rand() % pis.size(); 
```

Comment:
This uses the std::rand I believe, I would use a random object you already have or a new one in the same way the other ones created to control the random seeds.